### PR TITLE
Removed the focus ring when the focus is from the mouse

### PR DESCRIPTION
**Pull request by silviuaavram**
Removing focus ring when the focus is applied by mouse. Only added it on Buttons for now.

Checking with the library what-input the source of the focus. If focus comes from keyboard, make the state property `isLastFocusFromKeyboard` true, which triggers the custom style from the buttonStyles.ts

### CODE REVIEW

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -92,17 +94,26 @@ class Button extends UIComponent<any, any> {
...
 
-  public renderComponent({ ElementType, classes, accessibility, rest }): React.ReactNode {
+  public state = {
+    isLastFocusFromMouse: false,
+  }
+
+  public renderComponent({ ElementType, classes, accessibility, styles, rest }): React.ReactNode {
     const { children, content, disabled, iconPosition } = this.props
+    const { isLastFocusFromMouse } = this.state
+    console.log('state is', isLastFocusFromMouse)
```
Comment (miroslavstastny) [inline src/components/Button/Button.tsx]: remove `console.log` before merge

File: src/components/Button/Button.tsx [hunk 2]
```diff
@@ -92,17 +94,26 @@ class Button extends UIComponent<any, any> {
...
+
     const hasChildren = childrenExist(children)
 
     return (
       <ElementType
         className={classes.root}
+        styles={isLastFocusFromMouse ? { root: styles.focusFromMouse } : undefined}
```
Comment (miroslavstastny) [inline src/components/Button/Button.tsx]: Does `styles` attribute work for html elements? I thought it works for Components only.
For html element it results in `<button ... styles="[object Object]" ...>`

**Comment by miroslavstastny** [comment]
btw, why don't we pass `state` together with `props` to `componentStyles()`? We do it for accessibility.
@kuzhelov @levithomason

File: src/components/Button/Button.tsx (hunk 2 above)
Comment (silviuaavram) [reply src/components/Button/Button.tsx]: we were trying a workaround and yes, we figured that ElemenType is not going through the renderComponent.

File: src/components/Button/Button.tsx (hunk 1 above)
Comment (silviuaavram) [reply src/components/Button/Button.tsx]: still a long way to go. the only way to go forward is probably adding state to the styles, as we do with props.

**Comment by kuzhelov** [comment]
waits for functionality that would allow to pass component's state information to style functions

File: src/components/Button/Button.tsx [hunk 3]
```diff
@@ -85,6 +87,13 @@ class Button extends UIComponent<Extendable<IButtonProps>, any> {
      */
     onClick: PropTypes.func,
 
+    /**
+     * Called after user's click.
```
Comment (jurokapsiar) [inline src/components/Button/Button.tsx]: click -> focus

**Comment by mnajdova** [comment]
We have another open PR that handles the focus styles in the Button alongside with other styling improvements (https://github.com/stardust-ui/react/pull/165). Can you check if those are handling the same thing and whether we will still need the focus styling from the other PR? I am asking because that one is already mark as ready for review.

File: src/components/Button/Button.tsx [hunk 4]
```diff
@@ -179,6 +194,16 @@ class Button extends UIComponent<Extendable<IButtonProps>, any> {
       onClick(e, this.props)
     }
   }
+
+  private handleFocus = (e: React.SyntheticEvent) => {
+    const { onFocus } = this.props
+
+    this.setState({ isLastFocusFromKeyboard: whatInput.ask() === 'keyboard' })
+
+    if (onFocus) {
```
Comment (miroslavstastny) [inline src/components/Button/Button.tsx:203]: in other places instead of check&call there is `_.invoke()` used

File: package.json [hunk 5]
```diff
@@ -158,7 +158,8 @@
     "typescript": "^3.0.1",
     "webpack": "^3.5.4",
     "webpack-dev-middleware": "^1.12.0",
-    "webpack-hot-middleware": "^2.18.2"
+    "webpack-hot-middleware": "^2.18.2",
+    "what-input": "^5.1.1"
```
Comment (miroslavstastny) [inline package.json:162]: dependency, not devDependency

File: src/components/Button/Button.tsx [hunk 6]
```diff
@@ -179,6 +194,16 @@ class Button extends UIComponent<Extendable<IButtonProps>, any> {
       onClick(e, this.props)
     }
   }
+
+  private handleFocus = (e: React.SyntheticEvent) => {
+    const { onFocus } = this.props
+
+    this.setState({ isLastFocusFromKeyboard: whatInput.ask() === 'keyboard' })
```
Comment (jurokapsiar) [inline src/components/Button/Button.tsx:201]: this line contains specific logic that will probably be reused in other components. Let's move it to a util function somewhere in the /lib directory, something like

```
  this.setState(focusFromKeyboardState())
```

**Pull request closed without merging**